### PR TITLE
Bert Example

### PR DESCRIPTION
**Pull request by HCY-11**
Example using convert_to_analog to run BERT Transformer on SQuAD task.

### CODE REVIEW

File: examples/24_bert_on_squad.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    rpu_config.forward.inp_res = 1/dac_res
+    rpu_config.forward.out_res = 1/adc_res
+    rpu_config.forward.bound_management = BoundManagementType.NONE
+    rpu_config.noise_model = PCMLikeNoiseModel(g_max=g_max)
+    rpu_config.drift_compensation = GlobalDriftCompensation()
+    return rpu_config
+
+def create_model_and_tokenizer(rpu_config):
+    model = AutoModelForQuestionAnswering.from_pretrained("bert-large-uncased-whole-word-masking-finetuned-squad")
+    model = convert_to_analog(model, rpu_config)
```
[the remark this answers is not in the document]
Comment (kaoutar55) [reply examples/24_bert_on_squad.py]: Hi @HCY-11 can you fix the sign off issue. You can follow the steps as detailed here: https://github.com/IBM/aihwkit/pull/440/checks?check_run_id=9880689052

This needs to be fixed before I can merge the example.

[the remark this answers is not in the document]
Comment (HCY-11) [reply examples/24_bert_on_squad.py]: The latest commits should have given credit and documentation

[the remark this answers is not in the document]
Comment (kaoutar55) [reply examples/24_bert_on_squad.py]: Thanks @HCY-11 for your nice contribution! There is one item remaining. Please resolve the conflicts in the readme file.

**Review by HCY-11 — COMMENTED** [review]
Changes integrated

**Comment by kaoutar55** [comment]
Hi @HCY-11 can you add the missing libraries to requirements file to fix the dependencies issues in the build: 
Examples/24_bert_on_squad.py:26:0: E0401: Unable to import 'transformers.integrations' (import-error)
755examples/24_bert_on_squad.py:28:0: E0401: Unable to import 'transformers' (import-error)
756examples/24_bert_on_squad.py:38:0: E0401: Unable to import 'evaluate' (import-error)
757examples/24_bert_on_squad.py:39:0: E0401: Unable to import 'datasets' (import-error)
758examples/24_bert_on_squad.py:79:4: E0401: Unable to import 'wandb' (import-error)

**Review by kaoutar55 — APPROVED** [review]

File: examples/24_bert_on_squad.py [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+    model.eval()
+
+    metric = load("squad")
+
+    ground_truth = [{"id": ex["id"], "answers": ex["answers"]} for ex in squad["validation"]]
+
+    t_inference_list = [0.0] + np.logspace(0, np.log10(float(max_inference_time)), n_times).tolist()
+
+    for t_inference in t_inference_list:
+        if not ARGS.digital:
```
Comment (maljoras) [inline examples/24_bert_on_squad.py:479]: It makes no sense to run multiple inferences for a purely digital model

File: examples/README.md [hunk 3]
```diff
@@ -516,6 +516,222 @@ tiki-taka v2 algorithm that achieves better performance than SGD.
...
+PARSER.add_argument("-w", "--wandb",
+                    help="Add to use wandb",
+                    action="store_true")
+PARSER.add_argument("-n", "--noise",
+                    help="Weight noise",
+                    default=0.0175,
+                    type=float)
+```
+
+### AiMOS NPL Environment Setup
```
Comment (maljoras) [inline examples/README.md:547]: should this reference to Aimos be in the public repo? I don't think so

File: examples/README.md [hunk 4]
```diff
@@ -516,6 +516,222 @@ tiki-taka v2 algorithm that achieves better performance than SGD.
...
+#SBATCH -D /gpfs/u/home/AHWT/AHWT<user-name>/scratch
+#SBATCH -o joboutput.%J
+#SBATCH -e joberror.%J
+#SBATCH --time=06:00:00
+#SBATCH --gres=gpu:1
+#SBATCH --gpus=1
+
+source ~/.bashrc
+conda activate <env-name>
+srun --mem=32G python ~/barn/aihwkit/examples/24_bert_on_squad.py --noise 0.1 &
```
Comment (maljoras) [inline examples/README.md:591]: please remove all the aimos explanation from this public example

File: examples/README.md [hunk 5]
```diff
@@ -516,6 +516,222 @@ tiki-taka v2 algorithm that achieves better performance than SGD.
...
+
+Follow the [installation instructions](https://docs.cci.rpi.edu/software/Conda/) to install conda.
+Then, create an environment with `conda create --name <env-name>`,
+and run `conda activate <env-name>`.
+
+After activation, install the necessary
+[dependences](https://aihwkit.readthedocs.io/en/latest/advanced_install.html),
+and run the following commands:
+
+`conda install pytorch torchvision torchaudio pytorch-cuda=11.6 -c pytorch -c nvidia`
```
Comment (maljoras) [inline examples/README.md:567]: can we add these dependencoes to the setup.py so that one can just use something like  
`pip install aihwkit[bert]`
to install all dependencies as we did for the visualization etc. ?

File: examples/README.md [hunk 6]
```diff
@@ -516,6 +516,222 @@ tiki-taka v2 algorithm that achieves better performance than SGD.
...
+https://github.com/huggingface/notebooks/blob/main/examples/question_answering.ipynb
+
+The example loads a pre-trained BERT model trained on
+the SQuAD dataset. It then applies `convert_to_analog()`
+to examine the effects of `drift_analog_weights()` on inference performance at
+different weight noise levels. Tensorboard is used to display the SQuAD
+metrics evaluated using the model at various times after training completed.
+
+Commandline arguments can be used to control certain options
+```python
```
Comment (maljoras) [inline examples/README.md:649]: make an real use case example here rather than pasting the parser

File: requirements.txt [hunk 7]
```diff
@@ -9,3 +9,7 @@ scipy
 requests>=2.25,<3
 numpy>=1.19
 protobuf>=3.15.0,<4
+transformers
```
Comment (maljoras) [inline requirements.txt:12]: please remove all these additional from requirements.txt! We cannot make this a global dependency. Use the extra dependency in setup.py for that

**Review by maljoras — CHANGES_REQUESTED** [review]
Many thanks great work! 
Please remove the extra dependency and Aimos cluster information though. These do not belong into the public repo.

File: examples/24_bert_on_squad.py [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+    # Define RPU configuration and use it to create model and tokenizer
+    rpu_config = create_ideal_rpu_config() if ARGS.ideal else create_rpu_config(w_noise=ARGS.noise)
+    model = create_model(rpu_config)
+    squad, tokenized_data, eval_data = create_datasets()
+    optimizer = create_optimizer(model)
+    trainer, writer = make_trainer(model, optimizer, tokenized_data)
+
+    # Do hw-aware training
+    if ARGS.hwa_train:
+        trainer.train()
```
[the remark this answers is not in the document]
Comment (HCY-11) [reply examples/24_bert_on_squad.py:527]: The model has already been fine-tuned to SQUAD in digital. If hwa_train=False, I believe this means no hardware-aware training is done.
What do you mean by separate fine-tuning and inference. Are they not already separate steps?

[the remark this answers is not in the document]
Comment (kaoutar55) [reply examples/24_bert_on_squad.py:547]: Hi @HCY-11 do you load the fine tuned squad model when you are doing the inference in analog? The code only does the fine-tuning when it class the trainer.train(), which we should do only once in analog domain and save a checkpoint. Next time we want to try some other noise parameters for the RPU config, we don't want to restart the fine-tuning step, so we can just load the checkpoint of the fine-tuned model.

[the remark this answers is not in the document]
Comment (HCY-11) [reply examples/24_bert_on_squad.py:562]: Yes, the fine tuned squad model is loaded for inference. I added checkpointing so trainer.train() should only run when needed. I am running into memory errors with the trainer at the moment.

File: setup.py [hunk 9]
```diff
@@ -75,6 +75,12 @@ def get_long_description() -> str:
...
     zip_safe=False,
     extras_require={
-        'visualization': ['matplotlib>=3.0']
+        'visualization': ['matplotlib>=3.0'],
+        'bert': [
+            'transformers',
+            'evaluate',
+            'datasets',
+            'wandb'
+        ]
```
Comment (maljoras) [inline setup.py:84]: seems that "tensorboard" is also needed, please add dependency

File: .travis.yml [hunk 10]
```diff
@@ -39,6 +39,7 @@ job_compile_common: &job_compile_common
     - $PYTHON_EXECUTABLE -m pip install -r requirements.txt
     # Install the package in editable mode.
     - VERBOSE=1 $PYTHON_EXECUTABLE -m pip install -v -e ".[visualization]"
+    - VERBOSE=1 $PYTHON_EXECUTABLE -m pip install -v ".[bert]"
```
Comment (maljoras) [inline .travis.yml:42]: can you remove this from travis again? we do not test the examples in travis. Thanks

File: .travis.yml [hunk 11]
```diff
@@ -100,6 +101,7 @@ jobs:
       - $PYTHON_EXECUTABLE -m pip install -r requirements.txt
       # Install the package in editable mode.
       - VERBOSE=1 $PYTHON_EXECUTABLE -m pip install -v -e ".[visualization]"
+      - VERBOSE=1 $PYTHON_EXECUTABLE -m pip install -v ".[bert]"
```
Comment (maljoras) [inline .travis.yml:104]: remove this line again. Just leave travis as was before.

File: examples/24_bert_on_squad.py [hunk 12]
```diff
@@ -0,0 +1,572 @@
...
+
+def make_trainer(model, optimizer, tokenized_data):
+    """Create the Huggingface Trainer"""
+    training_args = TrainingArguments(
+        output_dir="./",
+        save_strategy="no",
+        per_device_train_batch_size=4,
+        per_device_eval_batch_size=4,
+        num_train_epochs=3,
+        weight_decay=0.01,
```
Comment (maljoras) [inline examples/24_bert_on_squad.py:466]: are you sure about the weight decay if 0.01 ? Seems very big to me

**Review by maljoras — CHANGES_REQUESTED** [review]
@HCY-11 many thanks looks almost ready!

only these changes: 
- restore the old travis, we do not test the examples in travis so no need to install [bert]
- it seems that 'tensorboard' is needed as well, please add it in setup.py

**Review by maljoras — APPROVED** [review]

**Review by maljoras — APPROVED** [review]
